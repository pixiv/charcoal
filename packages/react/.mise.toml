[tools]
pnpm = "10.17.1"
