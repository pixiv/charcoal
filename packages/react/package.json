{
  "name": "@charcoal-ui/react",
  "version": "4.0.0-beta.3",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs.js",
      "import": "./dist/index.esm.js",
      "default": "./dist/index.esm.js"
    },
    "./dist/index.css": {
      "import": "./dist/index.css",
      "require": "./dist/index.css"
    }
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@charcoal-ui/esbuild-plugin-styled-components": "^4.0.0-beta.3",
    "@react-types/switch": "^3.1.2",
    "@storybook/addon-actions": "^7.4.1",
    "@storybook/addon-knobs": "^7.0.2",
    "@storybook/addons": "^7.4.1",
    "@storybook/api": "^7.4.1",
    "@storybook/components": "^7.4.1",
    "@storybook/core-events": "^7.4.1",
    "@storybook/react": "^7.4.1",
    "@storybook/react-webpack5": "^7.4.1",
    "@storybook/theming": "^7.4.1",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.0",
    "@types/jest-axe": "^3.5.3",
    "@types/react": "^17.0.38",
    "@types/react-dom": "^17.0.11",
    "@types/react-router-dom": "^5.3.3",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-axe": "^5.0.1",
    "jest-styled-components": "^7.0.8",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.1",
    "rimraf": "^3.0.2",
    "storybook": "^7.4.1",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/icons": "^4.0.0-beta.3",
    "@charcoal-ui/styled": "^4.0.0-beta.3",
    "@charcoal-ui/theme": "^4.0.0-beta.3",
    "@charcoal-ui/utils": "^4.0.0-beta.3",
    "@react-aria/button": "^3.9.1",
    "@react-aria/checkbox": "^3.13.0",
    "@react-aria/dialog": "^3.5.10",
    "@react-aria/focus": "^3.16.0",
    "@react-aria/link": "^3.6.5",
    "@react-aria/overlays": "^3.20.0",
    "@react-aria/radio": "^3.10.0",
    "@react-aria/ssr": "^3.9.1",
    "@react-aria/switch": "^3.6.0",
    "@react-aria/textfield": "^3.14.1",
    "@react-aria/utils": "^3.23.0",
    "@react-aria/visually-hidden": "^3.8.8",
    "polished": "^4.1.4",
    "react-spring": "^9.0.0",
    "react-stately": "^3.26.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=17.0.0",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/react"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
