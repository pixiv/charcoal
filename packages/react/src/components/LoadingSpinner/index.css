.charcoal-loading-spinner {
  --charcoal-loading-spinner-size: 48px;
  --charcoal-loading-spinner-padding: 16px;

  box-sizing: content-box;
  margin: auto;
  padding: var(--charcoal-loading-spinner-padding);
  border-radius: 8px;
  font-size: var(--charcoal-loading-spinner-size);
  width: var(--charcoal-loading-spinner-size);
  height: var(--charcoal-loading-spinner-size);
  opacity: 0.84;
  color: var(--charcoal-color-icon-tertiary-default);
  background-color: var(--charcoal-color-background-default);
}
.charcoal-loading-spinner[data-transparent='true'] {
  background-color: rgba(0, 0, 0, 0);
}

@keyframes charcoal-loading-spinner-icon-scale-out {
  from {
    transform: scale(0);
    opacity: 1;
  }
  to {
    transform: scale(1);
    opacity: 0;
  }
}

.charcoal-loading-spinner-icon {
  width: 1em;
  height: 1em;
  border-radius: 1em;
  background-color: currentColor;
  animation: charcoal-loading-spinner-icon-scale-out 1s both ease-out;
  animation-iteration-count: infinite;
}

.charcoal-loading-spinner-icon[data-reset-animation] {
  animation: none;
}
.charcoal-loading-spinner-icon[data-once='true'] {
  animation-iteration-count: 1;
}
