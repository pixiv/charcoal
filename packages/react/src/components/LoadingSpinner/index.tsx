import React, { useImperativeHandle, useRef } from 'react'
import styled, { keyframes } from 'styled-components'
import { theme } from '../../styled'

export default function LoadingSpinner({
  size = 48,
  padding = 16,
  transparent = false,
}) {
  return (
    <LoadingSpinnerRoot size={size} padding={padding} transparent={transparent}>
      <LoadingSpinnerIcon />
    </LoadingSpinnerRoot>
  )
}

const LoadingSpinnerRoot = styled.div.attrs({ role: 'progressbar' })<{
  size: number
  padding: number
  transparent: boolean
}>`
  margin: auto;
  padding: ${(props) => props.padding}px;
  border-radius: 8px;
  font-size: ${(props) => props.size}px;
  width: ${(props) => props.size}px;
  height: ${(props) => props.size}px;
  opacity: 0.84;
  ${({ transparent }) =>
    theme((o) => [
      o.font.text4,
      transparent ? o.bg.transparent : o.bg.background1,
    ])}
`

const scaleout = keyframes`
  from {
    transform: scale(0);
    opacity: 1;
  }

  to {
    transform: scale(1);
    opacity: 0;
  }
`

const Icon = styled.div.attrs({ role: 'presentation' })<{ once: boolean }>`
  width: 1em;
  height: 1em;
  border-radius: 1em;
  background-color: currentColor;
  animation: ${scaleout} 1s both ease-out;
  animation-iteration-count: ${(p) => (p.once ? 1 : 'infinite')};

  &[data-reset-animation] {
    animation: none;
  }
`

interface Props {
  once?: boolean
}

export interface LoadingSpinnerIconHandler {
  restart(): void
}

export const LoadingSpinnerIcon = React.forwardRef<
  LoadingSpinnerIconHandler,
  Props
>(function LoadingSpinnerIcon({ once = false }, ref) {
  const iconRef = useRef<HTMLDivElement>(null)

  useImperativeHandle(ref, () => ({
    restart: () => {
      if (!iconRef.current) {
        return
      }
      iconRef.current.dataset.resetAnimation = 'true'
      // Force reflow hack!
      void iconRef.current.offsetWidth
      delete iconRef.current.dataset.resetAnimation
    },
  }))

  return <Icon ref={iconRef} once={once} />
})
