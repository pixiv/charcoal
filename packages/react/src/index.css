:root {
  /* charcoal custom css variables */
  --charcoal-color-custom-focus-effect: 0 0 0 4px rgba(0, 150, 250, 0.32);
  --charcoal-color-custom-error-effect: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
