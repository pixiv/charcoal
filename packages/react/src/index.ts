export {
  default as ComponentAbstraction,
  useComponentAbstraction,
  type LinkProps,
} from './core/ComponentAbstraction'
export { SSRProvider } from './core/SSRProvider'
export { default as Button, type ButtonProps } from './components/Button'
export {
  default as Clickable,
  type ClickableProps,
  type ClickableElement,
} from './components/Clickable'
export {
  default as IconButton,
  type IconButtonProps,
} from './components/IconButton'
export {
  default as Radio,
  type RadioProps,
  RadioGroup,
  type RadioGroupProps,
} from './components/Radio'
export {
  default as Select,
  type SelectProps,
  SelectGroup,
  type SelectGroupProps,
} from './components/Select'
export { default as Switch, type SwitchProps } from './components/Switch'
export {
  default as TextField,
  type TextFieldProps,
  type SingleLineTextFieldProps,
  type MultiLineTextFieldProps,
} from './components/TextField'
export { default as Icon, type IconProps } from './components/Icon'
export { default as Modal, type ModalProps } from './components/Modal'
export {
  default as LoadingSpinner,
  LoadingSpinnerIcon,
} from './components/LoadingSpinner'
export {
  default as SegmentedControl,
  type SegmentedControlProps,
} from './components/SegmentedControl'
export { default as Checkbox } from './components/Checkbox'
export { default as TagItem, type TagItemProps } from "./components/TagItem"
