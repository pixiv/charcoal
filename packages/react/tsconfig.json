{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "types": ["vitest/globals", "@testing-library/jest-dom"],
    "isolatedDeclarations": true,
  },
  "include": ["./src", "./docs", "./vitest.setup.ts"]
}
