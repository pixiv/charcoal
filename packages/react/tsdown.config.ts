import { defineConfig } from 'tsdown'
// import { styledComponentsPlugin } from '@charcoal-ui/esbuild-plugin-styled-components'
// import postcss from 'postcss'
// import fs from 'node:fs/promises'
// import path from 'node:path'

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['esm', 'cjs'],
  // outExtension({ format }) {
  //   return {
  //     js: `.${format}.js`,
  //   }
  // },
  // banner: {
  //   js: `"use client";`,
  // },
  target: 'esnext',
  sourcemap: true,
  dts:true,
  // minify:true,
  
  // esbuildPlugins: [
  //   styledComponentsPlugin,
  //   {
  //     name: 'at-layer-charcoal',
  //     setup(build) {
  //       // NOTE: this will be called twice for esm and cjs.
  //       build.onEnd(async (result) => {
  //         const originalCssOutput = result.outputFiles?.find((file) =>
  //           file.path.endsWith('css')
  //         )
  //         if (!originalCssOutput) {
  //           throw new Error('[at-layer-charcoal]: expect original css output')
  //         }

  //         const layeredCssFilePath = originalCssOutput.path.replace(
  //           'index.css',
  //           'layered.css'
  //         )
  //         const layeredCssOutput = await postcss({
  //           postcssPlugin: 'at-layer-charcoal',
  //           Once(_, { AtRule, result }) {
  //             const atLayerNode = new AtRule({
  //               name: 'layer',
  //               params: 'charcoal',
  //               nodes: result.root.nodes,
  //             })
  //             result.root.nodes = [atLayerNode]
  //           },
  //         }).process(originalCssOutput.text, {
  //           from: originalCssOutput.path,
  //           to: layeredCssFilePath,
  //           map: {
  //             inline: false,
  //           },
  //         })

  //         // https://github.com/evanw/esbuild/issues/2999
  //         // this will be called before dist is created
  //         await fs.mkdir(path.dirname(layeredCssFilePath), { recursive: true })
  //         await Promise.all([
  //           fs.writeFile(layeredCssFilePath, layeredCssOutput.content),
  //           fs.writeFile(
  //             `${layeredCssFilePath}.map`,
  //             layeredCssOutput.map.toString()
  //           ),
  //         ])

  //         // eslint-disable-next-line no-console
  //         console.log(
  //           `${build.initialOptions.format?.toUpperCase()} dist/layered.css`
  //         )
  //       })
  //     },
  //   },
  // ],
})
