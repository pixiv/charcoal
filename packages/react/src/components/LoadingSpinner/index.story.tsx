import { Meta, StoryObj } from '@storybook/react'
import LoadingSpinner from '.'

export default {
  title: 'LoadingSpinner',
  component: LoadingSpinner,
  tags: ['skip-test'],
  parameters: {
    layout: 'centered',
  },
} as Meta<typeof LoadingSpinner>

export const Default: StoryObj<typeof LoadingSpinner> = {
  render: (props) => <LoadingSpinner {...props} />,
}

export const Transparent: StoryObj<typeof LoadingSpinner> = {
  render: () => <LoadingSpinner transparent />,
}

export const Size: StoryObj<typeof LoadingSpinner> = {
  render: () => <LoadingSpinner size={128} />,
}

export const Padding: StoryObj<typeof LoadingSpinner> = {
  render: () => <LoadingSpinner padding={24} />,
}
