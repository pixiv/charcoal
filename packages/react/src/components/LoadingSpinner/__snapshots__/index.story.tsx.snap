// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots LoadingSpinner Basic 1`] = `
.c0 {
  box-sizing: content-box;
  margin: auto;
  padding: 16px;
  border-radius: 8px;
  font-size: 48px;
  width: 48px;
  height: 48px;
  opacity: 0.84;
  color: var(--charcoal-text4);
  background-color: var(--charcoal-background1);
}

.c1 {
  width: 1em;
  height: 1em;
  border-radius: 1em;
  background-color: currentColor;
  -webkit-animation: cNZOgq 1s both ease-out;
  animation: cNZOgq 1s both ease-out;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

.c1[data-reset-animation] {
  -webkit-animation: none;
  animation: none;
}

<div
  data-dark={false}
>
  <div
    className="c0 basic"
    role="progressbar"
    size={48}
  >
    <div
      className="c1"
      role="presentation"
    />
  </div>
</div>
`;

exports[`Storyshots LoadingSpinner Icon 1`] = `
.c0 {
  width: 1em;
  height: 1em;
  border-radius: 1em;
  background-color: currentColor;
  -webkit-animation: cNZOgq 1s both ease-out;
  animation: cNZOgq 1s both ease-out;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

.c0[data-reset-animation] {
  -webkit-animation: none;
  animation: none;
}

<div
  data-dark={false}
>
  <div
    css="
        font-size: 12px;
        color: #B1CC29;
      "
  >
    <div
      className="c0"
      role="presentation"
    />
  </div>
</div>
`;
